feat: Add Rockset as an OnlineStore

danielin917 opened the pull request:
This commit adds Rockset as a contributed online store. It implements the Update, Teardown, Read and Write apis using the Rockset python client.

Signed-off-by: Daniel Lin <dan@rockset.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. Ensure that your code follows our code conventions: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#code-style--linting
2. Run unit tests and ensure that they are passing: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#unit-tests
3. If your change introduces any API changes, make sure to update the integration tests here: https://github.com/feast-dev/feast/tree/master/sdk/python/tests
4. Make sure documentation is updated for your PR!
5. Make sure your commits are signed: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#signing-off-commits
6. Make sure your PR title follows conventional commits (e.g. fix: [description] vs feat: [description])

-->

**What this PR does / why we need it**:
This PR adds Rockset as an implemented OnlineFeatureStore

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #
N/A

danielin917 commented:
Hello! Integrations tests have all passed for the online store. Did not add a local universal test to the Makefile since there is no docker image to use.

danielin917 commented:
/assign @mavysavydav

danielin917 commented:
👋 Just checking in to see if there is anything I should do to make this easier to review

mavysavydav commented:
hi! Do you know who might have context on these changes who may be better suited to review? I also recently stepped down from my role as a maintainer so am not as involved in reviewing these for the time being. Apologies for the delay! If there's nobody you have in mind who might be a good fit for reviewing, I can pass this on to someone else in the maintainer team.

danielin917 commented:
Not sure who the best person is, I think @felixwang9817 mentioned he was interested but not sure who else could review

mavysavydav commented:
@felixwang9817 -- were you interested in taking a look at this or do you know who might be?

adchia commented on sdk/python/requirements/py3.9-ci-requirements.txt:
```diff
@@ -1,6 +1,6 @@
 #
-# This file is autogenerated by pip-compile with python 3.9
-# To update, run:
+# This file is autogenerated by pip-compile with Python 3.10
```
seems like you generated this with the wrong python version?

danielin917 replied:
Yea, thanks for the catch!

felixwang9817 commented on docs/SUMMARY.md:
```diff
@@ -89,13 +89,15 @@
   * [Overview](reference/online-stores/overview.md)
   * [SQLite](reference/online-stores/sqlite.md)
   * [Snowflake](reference/online-stores/snowflake.md)
+  * [Rockset](reference/online-stores/rockset.md)
```
this line can be removed - we only need the `[Rockset (contrib)]` line!

felixwang9817 commented on docs/reference/online-stores/README.md:
```diff
@@ -14,6 +14,10 @@ Please see [Online Store](../../getting-started/architecture-and-components/onli
 [snowflake.md](snowflake.md)
 {% endcontent-ref %}
 
+{% content-ref url="rockset.md" %}
```
similar to `SUMMARY.md`, this block can be removed

felixwang9817 commented on docs/reference/online-stores/rockset.md:
```diff
@@ -0,0 +1,36 @@
+# Rockset
```
would make this `Rockset (contrib)`

felixwang9817 commented on docs/reference/online-stores/rockset.md:
```diff
@@ -0,0 +1,36 @@
+
+## Description
+
+*Alpha Development
```
this is formatted strangely, would probably turn this into a sentence

felixwang9817 commented on docs/reference/online-stores/rockset.md:
```diff
@@ -0,0 +1,36 @@
+
+The [Rockset](https://rockset.com/demo-signup/) online store provides supports for materializing feature values within a Rockset collection for serving online features in real-time.
+
+* Each document is uniquely identified by it's _id value. Repeated inerts into the same document will result in an upsert.
```
it's -> its

inerts -> inserts

felixwang9817 commented on docs/reference/online-stores/rockset.md:
```diff
@@ -0,0 +1,36 @@
+
+Rockset indexes all columns allowing for quick per feature look up and also allows for a dynamic typed schema that can change based on any new requirements. ApiKeys can be found in the console
+along with host urls which you can find in "View Region Endpoint Urls".
+Data Model Used Per Doc
```
this section is formatted strangely, might turn this into a code block?

felixwang9817 commented on sdk/python/requirements/py3.8-ci-requirements.txt:
```diff
@@ -1,6 +1,6 @@
 #
-# This file is autogenerated by pip-compile with python 3.8
-# To update, run:
+# This file is autogenerated by pip-compile with Python 3.10
```
you should set up a virtualenv with python 3.8 and actually run this command with python 3.8

danielin917 replied:
Yup all done, I naively thought updating the local env var would do everything

felixwang9817 commented on setup.py:
```diff
@@ -196,6 +200,7 @@
     + HBASE_REQUIRED
     + CASSANDRA_REQUIRED
     + AZURE_REQUIRED
+    + ROCKSET_REQUIRED
```
you should also add this down below to `extras_require`

this will allow people to install Feast with Rockset by doing `pip install feast[rockset]`

felixwang9817 reviewed (commented):
looks good overall! just a few nits

felixwang9817 commented:
hey @danielin917 thanks for the PR and sorry for the delay on the review - I'm happy to help you get this PR merged

I left a few nits above, and I think you'll want to rebase to ensure you can pass the latest tests!

danielin917 commented:
Thanks for all the comments!

felixwang9817 reviewed (approved):
/lgtm

felixwang9817 commented:
@danielin917 I stamped but looks like there's actually a few outstanding errors that need to be addressed

felixwang9817 reviewed (approved):
/lgtm

felixwang9817 commented:
btw @danielin917 for the DCO check you'll need to make sure all your commits are signed off properly!

danielin917 commented:
@felixwang9817 A few changes since last lgtm to get everything working.
- The biggest is that I commented out Rockset from the integration repo config. Until we have a dedicated instance to run CI for this like https://github.com/feast-dev/feast/pull/2258/files we will not have an account to test against (not sure if this can be provided for free in the near future).
- Next I pinned mypy to 0.98.1 since the upgrade manifested a bunch of unrelated lints that might be better handled in a separate diff
- Also, I added an explicit dependency on httpx as it was recently changed for the Starlette dependency which is needed by fastapi and there is currently no great way to pull through fastapi itself: https://github.com/tiangolo/fastapi/issues/5656
- Finally a few lint edits for the new mypy version that I added before pinning due to the overwhelming number of changes needed in other files. (Rockset integration tests still passed when run against my account)

adchia commented:
Still seeing a lot of lint errors?

<img width="1325" alt="Screenshot 2023-01-23 at 1 15 35 PM" src="https://user-images.githubusercontent.com/1476382/214117818-45ae6e2a-9580-453e-9870-dfa609952d5e.png">

danielin917 commented:
> Still seeing a lot of lint errors?
> 
> <img alt="Screenshot 2023-01-23 at 1 15 35 PM" width="1325" src="https://user-images.githubusercontent.com/1476382/214117818-45ae6e2a-9580-453e-9870-dfa609952d5e.png">
Sorry for the back and forth. pylint errors fixed up and locally runs clean now. I had to pin types-protobuf to get rid of the lint issues introduced in 4+ lots of:
```
Library stubs not installed for "google.*"
```

adchia commented:
/lgtm

adchia commented:
This seems fine. We'll probably have to revisit once libraries like tensorflow force protobuf v4

danielin917 commented:
Hmm looks like bigquery test failed with rate limit error. Let mee see if I can retrigger

felixwang9817 reviewed (approved):
/lgtm

danielin917 commented:
ERROR: Cannot uninstall types-protobuf 4.21.0.3, RECORD file not found. You might be able to recover from this via: 'pip install --force-reinstall --no-deps types-protobuf==4.21.0.3'.

Not sure if there is a way to just rerun this test and this goes away or might need to play with dependancies more.

felixwang9817 commented:
> ERROR: Cannot uninstall types-protobuf 4.21.0.3, RECORD file not found. You might be able to recover from this via: 'pip install --force-reinstall --no-deps types-protobuf==4.21.0.3'.
> 
> Not sure if there is a way to just rerun this test and this goes away or might need to play with dependancies more.

@danielin917 looks like it's probably just a dependencies issue - I'll look into it today/tmrw!

felixwang9817 commented on setup.py:
```diff
@@ -151,7 +156,7 @@
         "mock==2.0.0",
         "moto<4",
-        "mypy>=0.931",
+        "mypy>=0.931, <=0.981",
```
@danielin917 why is this necessary?

danielin917 replied:
I was seeing a lot of unrelated lint errors before from different files. But actually now after removing the cap lint passed so not sure what changed

felixwang9817 commented on setup.py:
```diff
@@ -174,7 +179,7 @@
         "pip-tools",
         "pybindgen",
-        "types-protobuf",
+        "types-protobuf<4",
```
@danielin917 why is this necessary? the error message in the failing unit tests is `ERROR: Cannot uninstall types-protobuf 4.21.0.3, RECORD file not found. You might be able to recover from this via: 'pip install --force-reinstall --no-deps types-protobuf==4.21.0.3'.`, so this line seems very suspicious to me

danielin917 replied:
Hey yea, mentioned above that I pinned this type because the linter was producing a lot of errors using the new version `Library stubs not installed for "google.*"`. One thing that might be causing this is that we have:
 "mypy-protobuf==3.1",

danielin917 commented:
Not sure what happened between earlier testing and now but after removing cap on mypy and types-proto and this time I saw no lint errors locally

danielin917 commented:
Ugh, I am going to open a separate commit for upgrading mypy and deal with all of the linting issues there to unblock this diff as is instead of playing around with package dependencies

danielin917 commented:
Sorry to keep pinging just want to test the workflow after the rebase

felixwang9817 reviewed (approved):
/lgtm

felixwang9817 commented:
@danielin917 we finally got this merged in!! thanks a ton for the contribution, and for your patience throughout the review process - I know we ran into a bunch of issues throughout 😅

danielin917 commented:
Ahh sweet! Thanks @felixwang9817 and @adchia for the assist!
